source/cgal/surface_mesh.cc

```diff
@@ -0,0 +1,121 @@
+
+#include <deal.II/base/patterns.h>
+
+#include <deal.II/cgal/surface_mesh.h>
+
+#ifdef DEAL_II_WITH_CGAL
+
+DEAL_II_NAMESPACE_OPEN
+
+namespace
+{
+  template <typename dealiiFace, typename Container, typename CGAL_Mesh>
+  void
+  add_facet(const dealiiFace &face,
+            const Container & deal2cgal,
+            CGAL_Mesh &       mesh,
+            const bool        clockwise_ordering = true)
+  {
+    const unsigned                                nv = face->n_vertices();
+    std::vector<typename CGAL_Mesh::Vertex_index> indices;
+
+
+    switch (nv)
+      {
+        case 2:
+          mesh.add_edge(deal2cgal.at(face->vertex_index(0)),
+                        deal2cgal.at(face->vertex_index(1)));
+          break;
+        case 3:
+          indices = {deal2cgal.at(face->vertex_index(0)),
+                     deal2cgal.at(face->vertex_index(1)),
+                     deal2cgal.at(face->vertex_index(2))};
+          break;
+        case 4:
+          indices = {deal2cgal.at(face->vertex_index(0)),
+                     deal2cgal.at(face->vertex_index(1)),
+                     deal2cgal.at(face->vertex_index(3)),
+                     deal2cgal.at(face->vertex_index(2))};
+          break;
+        default:
+          Assert(false, ExcInternalError());
+          break;
+      }
+    auto f = mesh.null_face();
+    if (clockwise_ordering)
+      f = mesh.add_face(indices);
+    else
+      {
+        std::reverse(indices.begin(), indices.end());
+        f = mesh.add_face(indices);
+      }
+    Assert(f != mesh.null_face(),
+           ExcInternalError("While trying to build a CGAL facet, "
+                            "CGAL encountered a orientation problem that it "
+                            "was not able to solve."));
+  }
+} // namespace
+
+
+
+#  ifndef DOXYGEN
+// Template implementations
+namespace CGALWrappers
+{
+  template <typename CGALPointType, int dim, int spacedim>
+  void
+  to_cgal_mesh(const typename Triangulation<dim, spacedim>::cell_iterator &cell,
+               const Mapping<dim, spacedim> &     mapping,
+               CGAL::Surface_mesh<CGALPointType> &mesh)
+  {
+    Assert(dim > 1, ExcImpossibleInDim(dim));
+    using Mesh           = CGAL::Surface_mesh<CGALPointType>;
+    using Vertex_index   = typename Mesh::Vertex_index;
+    const auto &vertices = mapping.get_vertices(cell);
+
+    std::map<unsigned int, Vertex_index> deal2cgal;
+    // Add all vertices to the mesh
+    // Store CGAL ordering
+    for (const auto &i : cell->vertex_indices())
+      deal2cgal[cell->vertex_index(i)] =
+        mesh.add_vertex(CGALWrappers::to_cgal<CGALPointType>(vertices[i]));
+
+    // Add faces
+    if (dim < 3)
+      // simplices and quads
+      add_facet(cell, deal2cgal, mesh);
+    else
+      // faces of 3d cells
+      for (const auto &f : cell->face_indices())
+        add_facet(cell->face(f),
+                  deal2cgal,
+                  mesh,
+                  (f % 2 == 0 || cell->n_vertices() != 8));
```

Comment:
This logic looks odd. I guess the second part checks if the cell is a cube or not. The first part exploits that all edges in deal.II are orientated in increasing lex order?

Reply:
indeed this logic needs to be there only for hexes. In all other cases, we orient all faces consistently. In hexes, our orientation of the normal of the faces cannot be deduced by the ordering of the face vertices, but it depends on the index of the face. For even faces, 0,1,3,4 produces inward normal, for odd faces, it produces outward normal.

Reply:
Could you add this comment to the code. Furthermore, I would appreciate if you'd remove the `8` and replace it by the comparison with the reference element.